### PR TITLE
Rewrite hex encoder/decoder

### PR DESCRIPTION
**Pull request by Bond-009**

### CODE REVIEW

File: Emby.Dlna/Emby.Dlna.csproj [hunk 1]
```diff
@@ -12,7 +12,7 @@
   </ItemGroup>
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
Comment (JustAMan) [inline Emby.Dlna/Emby.Dlna.csproj]: why?

**Comment by Bond-009** [comment]
@JustAMan This wasn't solely about speed, it's also about a cleaner interface, more flexibility and more tests.

File: Emby.Dlna/Emby.Dlna.csproj (hunk 1 above)
Comment (Bond-009) [reply Emby.Dlna/Emby.Dlna.csproj]: Needed for `ReadOnlySpan`

**Comment by Bond-009** [comment]
Encode benches:
<table>
<thead><tr><th>                   Method</th><th>N</th><th>       Mean</th><th>    Error</th><th>     StdDev</th><th>Gen 0</th><th>Gen 1</th><th>Gen 2</th><th>Allocated</th>
</tr>
</thead><tbody><tr><td>HexEncode</td><td>0</td><td>14.957 ns</td><td>0.2963 ns</td><td>0.2772 ns</td><td>0.0076</td><td>-</td><td>-</td><td>24 B</td>
</tr><tr><td>BitConverterToString</td><td>0</td><td>2.979 ns</td><td>0.1226 ns</td><td>0.1147 ns</td><td>-</td><td>-</td><td>-</td><td>-</td>
</tr><tr><td>BitConverterToStringWithReplace</td><td>0</td><td>19.818 ns</td><td>0.4510 ns</td><td>0.4219 ns</td><td>-</td><td>-</td><td>-</td><td>-</td>
</tr><tr><td>HexEncode</td><td>10</td><td>47.225 ns</td><td>0.4554 ns</td><td>0.3803 ns</td><td>0.0408</td><td>-</td><td>-</td><td>128 B</td>
</tr><tr><td>BitConverterToString</td><td>10</td><td>48.904 ns</td><td>0.4388 ns</td><td>0.4105 ns</td><td>0.0255</td><td>-</td><td>-</td><td>80 B</td>
</tr><tr><td>BitConverterToStringWithReplace</td><td>10</td><td>258.847 ns</td><td>1.7845 ns</td><td>1.5819 ns</td><td>0.0458</td><td>-</td><td>-</td><td>144 B</td>
</tr><tr><td>HexEncode</td><td>100</td><td>278.936 ns</td><td>1.9976 ns</td><td>1.8686 ns</td><td>0.2699</td><td>-</td><td>-</td><td>848 B</td>
</tr><tr><td>BitConverterToString</td><td>100</td><td>364.007 ns</td><td>6.8167 ns</td><td>6.3763 ns</td><td>0.1988</td><td>-</td><td>-</td><td>624 B</td>
</tr><tr><td>BitConverterToStringWithReplace</td><td>100</td><td>2,167.155 ns</td><td>34.1497 ns</td><td>30.2728 ns</td><td>0.3319</td><td>-</td><td>-</td><td>1048 B</td>
</tr><tr><td>HexEncode</td><td>1000</td><td>2,556.766 ns</td><td>49.1590 ns</td><td>43.5782 ns</td><td>2.5635</td><td>-</td><td>-</td><td>8048 B</td>
</tr><tr><td>BitConverterToString</td><td>1000</td><td>3,147.860 ns</td><td>49.1378 ns</td><td>43.5594 ns</td><td>1.9188</td><td>-</td><td>-</td><td>6024 B</td>
</tr><tr><td>BitConverterToStringWithReplace</td><td>1000</td><td>21,154.019 ns</td><td>271.3156 ns</td><td>253.7888 ns</td><td>3.1738</td><td>-</td><td>-</td><td>10048 B</td>
</tr><tr><td>HexEncode</td><td>10000</td><td>26,133.436 ns</td><td>509.9760 ns</td><td>477.0319 ns</td><td>25.2991</td><td>-</td><td>-</td><td>80048 B</td>
</tr><tr><td>BitConverterToString</td><td>10000</td><td>31,219.015 ns</td><td>602.4661 ns</td><td>591.7023 ns</td><td>18.8599</td><td>-</td><td>-</td><td>60024 B</td>
</tr><tr><td>BitConverterToStringWithReplace</td><td>10000</td><td>212,505.170 ns</td><td>2,518.7480 ns</td><td>2,103.2687 ns</td><td>31.0059</td><td>-</td><td>-</td><td>100080 B</td>
</tr><tr><td>HexEncode</td><td>1000000</td><td>4,546,217.405 ns</td><td>89,573.3874 ns</td><td>95,842.5210 ns</td><td>1960.9375</td><td>1960.9375</td><td>320.3125</td><td>4000024 B</td>
</tr><tr><td>BitConverterToString</td><td>1000000</td><td>4,375,730.269 ns</td><td>230,531.3405 ns</td><td>679,726.8588 ns</td><td>343.7500</td><td>273.4375</td><td>234.3750</td><td>6000024 B</td>
</tr><tr><td>BitConverterToStringWithReplace</td><td>1000000</td><td>24,210,962.962 ns</td><td>533,714.0208 ns</td><td>1,114,059.9110 ns</td><td>1000.0000</td><td>750.0000</td><td>593.7500</td><td>10076209 B</td>
</tr></tbody></table>

Decode benches:
<table>
<thead><tr><th>   Method</th><th>N</th><th>       Mean</th><th>    Error</th><th>   StdDev</th><th>Gen 0</th><th>Gen 1</th><th>Gen 2</th><th>Allocated</th>
</tr>
</thead><tbody><tr><td>Decode</td><td>0</td><td>6.955 ns</td><td>0.2017 ns</td><td>0.1887 ns</td><td>-</td><td>-</td><td>-</td><td>-</td>
</tr><tr><td>DecodeSubString</td><td>0</td><td>7.228 ns</td><td>0.2396 ns</td><td>0.4729 ns</td><td>0.0076</td><td>-</td><td>-</td><td>24 B</td>
</tr><tr><td>Decode</td><td>10</td><td>35.780 ns</td><td>0.6645 ns</td><td>0.5890 ns</td><td>0.0127</td><td>-</td><td>-</td><td>40 B</td>
</tr><tr><td>DecodeSubString</td><td>10</td><td>306.395 ns</td><td>2.6608 ns</td><td>2.4889 ns</td><td>0.1144</td><td>-</td><td>-</td><td>360 B</td>
</tr><tr><td>Decode</td><td>100</td><td>256.257 ns</td><td>5.2144 ns</td><td>4.8776 ns</td><td>0.0405</td><td>-</td><td>-</td><td>128 B</td>
</tr><tr><td>DecodeSubString</td><td>100</td><td>2,943.328 ns</td><td>22.6101 ns</td><td>21.1495 ns</td><td>1.0605</td><td>-</td><td>-</td><td>3328 B</td>
</tr><tr><td>Decode</td><td>1000</td><td>2,227.279 ns</td><td>26.2846 ns</td><td>24.5867 ns</td><td>0.3242</td><td>-</td><td>-</td><td>1024 B</td>
</tr><tr><td>DecodeSubString</td><td>1000</td><td>28,944.279 ns</td><td>564.6135 ns</td><td>528.1398 ns</td><td>10.4980</td><td>-</td><td>-</td><td>33024 B</td>
</tr><tr><td>Decode</td><td>10000</td><td>21,533.739 ns</td><td>138.6735 ns</td><td>115.7986 ns</td><td>3.1738</td><td>-</td><td>-</td><td>10024 B</td>
</tr><tr><td>DecodeSubString</td><td>10000</td><td>300,002.720 ns</td><td>5,865.5232 ns</td><td>9,637.2228 ns</td><td>104.9805</td><td>-</td><td>-</td><td>330024 B</td>
</tr><tr><td>Decode</td><td>1000000</td><td>2,393,423.003 ns</td><td>36,877.0320 ns</td><td>30,793.9919 ns</td><td>171.8750</td><td>171.8750</td><td>152.3438</td><td>-</td>
</tr><tr><td>DecodeSubString</td><td>1000000</td><td>30,313,660.512 ns</td><td>593,480.5604 ns</td><td>609,460.8407 ns</td><td>10281.2500</td><td>93.7500</td><td>93.7500</td><td>32000001 B</td>
</tr></tbody></table>

**Pull request merged**